source/util/parse_number.h

```diff
@@ -0,0 +1,143 @@
+}
+
+// Tests whether any member of the IdTypes do not match.
+inline bool operator!=(const IdType& first, const IdType& second) {
+  return !(first == second);
+}
+
+// Returns true if the type is a scalar integer type.
+inline bool isScalarIntegral(const IdType& type) {
+  return type.type_class == IdTypeClass::kScalarIntegerType;
+}
+
+// Returns true if the type is a scalar floating point type.
+inline bool isScalarFloating(const IdType& type) {
+  return type.type_class == IdTypeClass::kScalarFloatType;
+}
+
+// Returns the number of bits in the type.
+// This is only valid for bottom, scalar integer, and scalar floating
+// classes.  For bottom, assume 32 bits.
+inline int assumedBitWidth(const IdType& type) {
+  switch (type.type_class) {
+    case IdTypeClass::kBottom:
+      return 32;
+    case IdTypeClass::kScalarIntegerType:
+    case IdTypeClass::kScalarFloatType:
+      return type.bitwidth;
+    default:
+      break;
+  }
+  // We don't care about this case.
+  return 0;
+}
+
+// Enum to indicate the parsing and encoding status.
+enum class EncodeNumberStatus {
+  kSuccess = 0,
+  // Unsupported bit width etc.
+  kUnsupported,
+  // Expected type (IdType) is not a scalar int or float, or putting a
+  // negative number in an unsigned literal.
+  kInvalidUsage,
+  // Number value does not fit the bit width of the expected type etc.
+  kInvalidText,
+};
+
+// Parses a numeric value of a given type from the given text.  The number
+// should take up the entire string, and should be within bounds for the
+// target type.  On success, returns true and populates the object
+// referenced by value_pointer. On failure, returns false.
+template <typename T>
+bool ParseNumber(const char* text, T* value_pointer);
+
+// Parses an integer value of a given |type| from the given |text| and encodes
+// the number by the given |emit| function. On success, returns
+// EncodeNumberStatus::kSuccess and an empty error message string, the parsed
+// number will be consumed by the given |emit| function word by word (least
+// significant word first). On failure, this function returns the error
+// encoding status and error message string to describe the error, |emit|
+// function will not be called on failure. Integers up to 64 bits are
+// supported.
+std::tuple<EncodeNumberStatus, std::string> ParseAndEncodeIntegerNumber(
+    const char* text, const IdType& type, std::function<void(uint32_t)> emit);
+
+// Parses a floating point value of a given |type| from the given |text| and
+// encodes the number by the given |emit| funciton. On success, returns
+// EncodeNumberStatus::kSuccess and an empty error message string, the parsed
+// number will be consumed by the given |emit| function word by word (least
+// significant word first). On failure, this function returns the error
+// encoding status and error message string to describe the error, |emit|
+// function will not be called on failure. Only 32 and 64 bit floating point
```

Comment:
I'm confused here too. The original comment shows only 32 and 64 bit are supported. But we do have tests for 16 bit float. I guess 16 bit floating point is also supported by now.

I should have a comment here for code review before.